### ISSUE TITLE
Comparison of two nodes fails

### ISSUE THREAD

**Issue by matjobst** [neuromorphs/NIR#36]
The comparison of two nodes fails: 
```python 
a = nir.Input(shape=np.array([1]))
b = nir.Input(shape=np.array([1]))
a == b   # "ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()"
```

The reason for that is that dataclasses generate a default comparison __eq__() function which compares all attributes of the class, but does not understand how to properly compare numpy arrays. For the case of the Input node, the resulting __eq__() does something like this:
```python
def __eq__(self, other):
    return self.shape == other.shape
```

Since the .shape is typically an array, it would rather need to do something like:

```python
def __eq__(self, other):
    return (self.shape == other.shape).all()
```

However I am still not sure whether we want that behavior. Because that means that:

```python
a = nir.Input(shape=np.array([1]))
b = nir.Input(shape=np.array([1]))
a == b   # True
```

This could be avoided if each node had its name also as an attribute.

The complete solution could be:
```python
@dataclass(eq=False)
class NIRNode:
    """Base superclass of Neural Intermediate Representation Unit (NIR).

    All NIR primitives inherit from this class, but NIRNodes should never be
    instantiated.
    """
    name: str
    def __eq__(self, other):
        print(self, other)
        if type(self) != type(other):
            return False
        for name,field in self.__dict__.items():
            print(name,field)
            if name not in other.__dict__.keys():
                return False
            other_field = other.__dict__[name]
            print(other_field)
            if type(field) != type(other_field):
                return False
            if isinstance(field, np.ndarray):
                if field.shape != other_field.shape or field.dtype != other_field.dtype:
                    return False
                if not (field == other_field).all():
                    return False
            else:
                if field != other_field:
                    return False
        return True
```

Since this is a breaking change (all nodes now have a required name attribute), and requires changing all tests, I will only create the pull request if everyone agrees on this type of solution.

**Comment by sheiksadique** [comment]
In the NIRTorch implementation of Graph and Nodes each node has a `name` attribute attached to it. 

But the equality of two nodes is determined by the element/torch module it holds which makes it easier to deal with.

**Comment by matjobst** [comment]
In the end it boils down to what equality should mean for nodes. Are two nodes equal when all their attributes are the same? When they have the same connectivity? Or if they are the identical object?

In my opinion, we should use the identity, meaning they are the same object. That prevents this:

```python
a = nir.Input(shape=np.array([1]))
b = nir.Input(shape=np.array([1]))
a == b   # True
```

The implementation of using identity for comparisons would be very simple, just overwriting the default __eq__ function:

```python
@dataclass(eq=False)
class NIRNode:
    """Base superclass of Neural Intermediate Representation Unit (NIR).

    All NIR primitives inherit from this class, but NIRNodes should never be
    instantiated.
    """
    def __eq__(self, other):
        return self is other
```

With this, the following works as I would expect:

```python
a = nir.Input(shape=np.array([1]))
b = nir.Input(shape=np.array([1]))
c = nir.Output(shape=np.array([1]))
a == a   # True
a == b   # False
a == c   # False
```

**Comment by Jegp** [comment]
NIR is explicitly declarative, meaning that a NIR node *declares* a type of computation. In that sense, I would actually argue that two NIR nodes, doing the same thing *should* be equal.

That said, I also agree that we two nodes in a graph should *not* be equal. But, the simplest way to achieve that would be to consider a "node in a graph" differently from simply a "node" in that nodes in a specific graph have names. So the comparison would really be (name1, node1) == (name2, node2).

Would you agree with that distinction @matjobst?

If so, the only remaining question is how to deal with the implementation. I can imagine two ways to address that
1. Allowing a == b to be true, and relying on the position/label in the graph to test for object differences
2. Adding the name attribute in to each `NIRNode`, in which case equality solves the problem since a == b will automatically check for the name

**Comment by sheiksadique** [comment]
Where do we need to compare two nodes for equality in practice? We will need to answer this to be able to clearly define equality operation. In general I agree with what @Jegp

**Comment by DylanMuir** [comment]
I can think of a case where I want to apply a transformation to a particular node in the graph, and I pass in the node and the graph to iterate over. Then we need to search for that node in the graph, and find *only identically that node*, not some other nodes with the same transfer function, or some other nodes which by chance or mistake have the same name. So nodes should have a unique ID which cannot be accidentally duplicated, and I should be able to use that ID when checking for equality.In practice most nodes in most networks will be automatically named, so I assume name collisions will happen often. On 18 Aug 2023, at 11:22 pm, Sadique Sheik ***@***.***> wrote:﻿
Where do we need to compare two nodes for equality in practice? We will need to answer this to be able to clearly define equality operation. In general I agree with what @Jegp

**Comment by matjobst** [comment]
@Jegp I agree with this distinction. I currently can not think of a use case where you would need the first type of equality.
For the second one, I agree with @DylanMuir and would add that a simple statement such as "node in graph" requires equality checks according to the second equality type.
The implementation of the second type does not actually require us to use IDs, see my previous suggestion:
> The implementation of using identity for comparisons would be very simple, just overwriting the default **eq** function:
> 
> ```python
> @dataclass(eq=False)
> class NIRNode:
>     """Base superclass of Neural Intermediate Representation Unit (NIR).
> 
>     All NIR primitives inherit from this class, but NIRNodes should never be
>     instantiated.
>     """
>     def __eq__(self, other):
>         return self is other
> ```

**Comment by Jegp** [comment]
I see the point, and I agree that we should expect collisions to happen often. @matjobst would you be up for giving this a go? Perhaps in #30 ?

**Comment by stevenabreu7** [comment]
closed with #30
